python/flow_sdk/shim_airbyte_cdk.py

```diff
@@ -165,22 +203,38 @@ def open(self, open: request.Open, emit: t.Callable[[Response], None]) -> None:
 
         catalog = ConfiguredAirbyteCatalog(streams=streams)
         config = open["capture"]["config"]
+        state = State(**(open["state"] or {}))
 
         emit(Response(opened=response.Opened(explicitAcknowledgements=False)))
 
-        for message in self.delegate.read(logger, config, catalog):
+        for message in self.delegate.read(
+            logger, config, catalog, state.to_airbyte_input()
+        ):
             if record := message.record:
                 entry = index[(record.namespace, record.stream)]
                 record.data.setdefault("_meta", {})["row_id"] = entry[1]
                 entry[1] += 1
 
-                emit(Response(
-                    captured=response.Captured(binding=entry[0], doc=record.data)
-                ))
+                emit(
+                    Response(
+                        captured=response.Captured(binding=entry[0], doc=record.data)
+                    )
+                )
 
-            elif state := message.state:
-                # TODO(johnny): handle various state types, and mix in next_row_id.
-                raise NotImplemented
+            elif state_msg := message.state:
+                state_msg = t.cast(AirbyteStateMessage, state_msg)
+                logger.info(f"Got a state message: {str(state_msg)}")
+                # TODO(johnny): mix in next_row_id
```

Comment:
So the problem as I understand it is that existing full refresh connectors don't remove stale rows from materializations because once the row is removed from the source system, there's fundamentally no positive signal that a materialization can react to that says "delete this row", just the negative signal that "the row doesn't exist in the next full refresh", which isn't formalized as a thing materializations know about. Right?

My understanding of your suggestion here would solve a subset of that problem, where a range of rows is deleted from the end of the source dataset, right? In that case, the capture connector would be able to say "ok, I went to X before but now I'm only seeing Y rows (where Y < X), so I can confidently say that rows X+1 through Y are deleted, so emit a positive deletion for those rows".

That seems... I guess good in that it's forward progress, but is it worth it vs a solution that solves the problem of emitting positive deletions for any row not seen in subsequent full refreshes? Maybe the problem is that that would require state to represent the key-space of previously-seen documents?

Reply:
>  is it worth it vs a solution that solves the problem of emitting positive deletions for any row not seen in subsequent full refreshes? Maybe the problem is that that would require state to represent the key-space of previously-seen documents?

Yes the problem is state. In principal we could track every key that is currently active, and upon a full refresh we could mark keys which are refreshed and then sweep to identify keys that were omitted.

This doesn't scale. There can be a truly unbounded number of keys. Even tracking 10M 32-byte keys, which is on the _very small side_ of tons of production use cases, is 320MB minimum before any kind of indexing overhead.

In comparison, tracking a single row offset is extremely cheap and works for any sized data.

> a solution that solves the problem of emitting positive deletions for any row not seen in subsequent full refreshes

I'll point out that keying on row offset in this way _is_ a solution that emits positive deletions for any rows not seen in subsequent full refreshes.

To be precise, all that we're not supporting is doing this for _arbitrary_ collection keys. But, in the garden path capture => materialize case, there's still a 1:1 correspondence between row offset and what the user considers to be their key, and it's easy to add an index.

For everything else, there are derivations.